src/test/ui/array-slice-vec/subslice-patterns-const-eval-match.rs

```diff
@@ -0,0 +1,110 @@
+// This test comprehensively checks the passing static and dynamic semantics
+// of subslice patterns `..`, `x @ ..`, `ref x @ ..`, and `ref mut @ ..`
+// in slice patterns `[$($pat), $(,)?]` .
+
+// run-pass
+
+#![feature(slice_patterns, const_fn, const_if_match)]
+#[derive(PartialEq, Debug, Clone)]
+struct N(u8);
+
+#[derive(PartialEq, Debug, Clone)]
+struct Z;
+
+macro_rules! n {
+    ($($e:expr),* $(,)?) => {
+        [$(N($e)),*]
+    }
+}
+
+macro_rules! zed {
+    ($e:expr) => { Z }
+}
+
+macro_rules! z {
+    ($($e:expr),* $(,)?) => {
+        [$(zed!($e)),*]
+    }
+}
+
+macro_rules! compare_evaluation {
+    ($e:expr, $t:ty $(,)?) => {{
+        const CONST_EVAL: $t = $e;
+        const fn const_eval() -> $t { $e }
+        const CONST_EVAL2: $t = const_eval();
+        assert_eq!(CONST_EVAL, $e);
+        assert_eq!(CONST_EVAL2, $e);
+    }}
+}
+
+macro_rules! repeat {
+    ($dollar:tt $mac:ident $tyname:ident; $($t:tt)*) => {
+        macro_rules! single {
+            ($dollar $mac:ident, $dollar $tyname:ident) => { $($t)* }
+        }
+        single!(n, N);
+        single!(z, Z);
+    }
+}
+
+fn main() {
+    repeat! {
+        $ arr Ty;
+        compare_evaluation!(
+            if let [_, x @ .., _] = &$arr!(1, 2, 3, 4) as &[_] { Some(x) } else { None },
```

Comment:
`if let` might not desugar to a `match` in the future, so it would be good to use `match` explicitly in some of these cases to future-proof things. (same in next test)